File: Nori.Server/Protocol/Messages.cs
// ────── ╔╗                                                                                    WGL
// ╔═╦╦═╦╦╬╣ Messages.cs
// ║║║║╬║╔╣║ Protocol message types for server-client communication
// ╚╩═╩═╩╝╚╝ ───────────────────────────────────────────────────────────────────────────────────────
using MessagePack;

namespace Nori;

// Enums ══════════════════════════════════════════════════════════════════════════

/// <summary>Discriminator for all server-to-client wire messages</summary>
public enum EMsgType : byte {
   SceneInit = 1,
   EntityAdd = 2,
   EntityRemove = 3,
   EntityUpdate = 4,
   PickResult = 5,
}

/// <summary>Scene type discriminator</summary>
public enum ESceneType : byte { Scene2D = 1, Scene3D = 2 }

/// <summary>Render primitive type discriminator</summary>
public enum EPrimType : byte {
   Lines2D = 1, Lines3D = 2, Beziers2D = 3,
   Points2D = 4, Points3D = 5,
   Mesh3D = 6,
   Text2D = 7, Text3D = 8, TextPx = 9,
   Fill2D = 10, Triangles2D = 11, Quads2D = 12,
}

/// <summary>Discriminator for client-to-server messages</summary>
public enum EClientMsgType : byte {
   ViewState = 10,
   Pick = 11,
   Interaction = 12,
   Command = 13,
}

/// <summary>Interaction event type</summary>
public enum EInteractionType : byte {
   Click = 1, DoubleClick = 2, Hover = 3,
   DragStart = 4, DragMove = 5, DragEnd = 6,
}

// Server → Client messages ═══════════════════════════════════════════════════════

/// <summary>Full scene snapshot sent on client connect</summary>
[MessagePackObject]
public class SceneInitMsg {
   [Key (0)] public ESceneType SceneType { get; set; }
   [Key (1)] public byte[] BgColor { get; set; } = [];    // RGBA bytes
   [Key (2)] public float[] Bounds { get; set; } = [];     // [minX,minY,maxX,maxY] or [minX,minY,minZ,maxX,maxY,maxZ]
   [Key (3)] public float[] Transforms { get; set; } = []; // Flattened Mat4F array (16 floats per transform)
   [Key (4)] public EntityDataMsg[] Entities { get; set; } = [];
}

/// <summary>One entity's complete render data</summary>
[MessagePackObject]
public class EntityDataMsg {
   [Key (0)] public int Id { get; set; }
   [Key (1)] public RenderPrimitive[] Primitives { get; set; } = [];
}

/// <summary>A single renderable primitive with its draw attributes</summary>
/// Vertex data is flattened floats whose interpretation depends on Type:
/// Lines2D/Points2D/Beziers2D/Fill2D/Triangles2D/Quads2D: pairs of (x,y)
/// Lines3D/Points3D: triples of (x,y,z)
/// Mesh3D: interleaved position+normal (x,y,z,nx,ny,nz per vertex)
/// Text2D: single (x,y) position
/// Text3D: single (x,y,z) position
/// TextPx: single (x,y) pixel position
[MessagePackObject]
public class RenderPrimitive {
   [Key (0)] public EPrimType Type { get; set; }
   [Key (1)] public float[] Data { get; set; } = [];                     // Vertex data (floats)
   [Key (2)] public int[]? Indices { get; set; }                         // Index data (optional, for Mesh3D triangles)
   [Key (3)] public byte[] Color { get; set; } = [255, 255, 255, 255];   // RGBA
   [Key (4)] public float LineWidth { get; set; } = 2f;
   [Key (5)] public byte LineType { get; set; }                          // ELineType cast to byte
   [Key (6)] public float LTScale { get; set; } = 30f;
   [Key (7)] public float PointSize { get; set; } = 4f;
   [Key (8)] public int TransformIndex { get; set; }
   [Key (9)] public int ZLevel { get; set; }
   [Key (10)] public byte ShadeMode { get; set; }                        // EShadeMode cast to byte
   // Text primitives
   [Key (11)] public string? Text { get; set; }
   [Key (12)] public byte TextAlign { get; set; }                        // ETextAlign cast to byte
   // Mesh primitives (wire edge indices separate from triangle indices)
   [Key (13)] public int[]? WireIndices { get; set; }
   // Fill primitives
   [Key (14)] public float[]? BoundData { get; set; }                    // [minX, minY, maxX, maxY]
}

/// <summary>Entity added to scene</summary>
[MessagePackObject]
public class EntityAddMsg {
   [Key (0)] public EntityDataMsg Entity { get; set; } = new ();
}

/// <summary>Entity removed from scene</summary>
[MessagePackObject]
public class EntityRemoveMsg {
   [Key (0)] public int EntityId { get; set; }
}

/// <summary>Entity updated (full replacement of render data)</summary>
[MessagePackObject]
public class EntityUpdateMsg {
   [Key (0)] public EntityDataMsg Entity { get; set; } = new ();
}

/// <summary>Result of a pick operation</summary>
[MessagePackObject]
public class PickResultMsg {
   [Key (0)] public int EntityId { get; set; }       // -1 if nothing picked
   [Key (1)] public float[] Position { get; set; } = []; // World position [x,y] or [x,y,z]
}

// Client → Server messages ═══════════════════════════════════════════════════════

/// <summary>Client's current view state (viewport, zoom, pan, rotation)</summary>
[MessagePackObject]
public class ViewStateMsg {
   [Key (0)] public int ViewportW { get; set; }
   [Key (1)] public int ViewportH { get; set; }
   [Key (2)] public double Zoom { get; set; } = 1;
   [Key (3)] public float PanX { get; set; }
   [Key (4)] public float PanY { get; set; }
   [Key (5)] public double XRot { get; set; }   // For 3D scenes
   [Key (6)] public double ZRot { get; set; }   // For 3D scenes
}

/// <summary>Pick request from client (screen coordinates)</summary>
[MessagePackObject]
public class PickMsg {
   [Key (0)] public int X { get; set; }
   [Key (1)] public int Y { get; set; }
}

/// <summary>Interaction event from client</summary>
[MessagePackObject]
public class InteractionMsg {
   [Key (0)] public EInteractionType Type { get; set; }
   [Key (1)] public float X { get; set; }         // World coordinates
   [Key (2)] public float Y { get; set; }
   [Key (3)] public float Z { get; set; }
   [Key (4)] public int Modifiers { get; set; }    // Shift=1, Ctrl=2, Alt=4
}

/// <summary>Command from client (e.g. switch demo, toggle option)</summary>
[MessagePackObject]
public class CommandMsg {
   [Key (0)] public string Name { get; set; } = "";
   [Key (1)] public string Arg { get; set; } = "";
}

File: Nori.Client/src/protocol/messages.ts
// protocol/messages.ts -- TypeScript message types mirroring C# Nori.Server.Protocol.Messages
// MessagePack with [Key(n)] attributes serializes as arrays where index = key number.

// ═══════════════════════════════════════════════════════════════════════════════
// Server → Client message types (EMsgType)
// ═══════════════════════════════════════════════════════════════════════════════
export const MsgType = {
  SceneInit: 1,
  EntityAdd: 2,
  EntityRemove: 3,
  EntityUpdate: 4,
  PickResult: 5,
} as const;

// ═══════════════════════════════════════════════════════════════════════════════
// Client → Server message types (EClientMsgType)
// ═══════════════════════════════════════════════════════════════════════════════
export const ClientMsgType = {
  ViewState: 10,
  Pick: 11,
  Interaction: 12,
  Command: 13,
} as const;

// ═══════════════════════════════════════════════════════════════════════════════
// Interaction types (EInteractionType)
// ═══════════════════════════════════════════════════════════════════════════════
export const InteractionType = {
  Click: 1,
  DoubleClick: 2,
  Hover: 3,
  DragStart: 4,
  DragMove: 5,
  DragEnd: 6,
} as const;

// ═══════════════════════════════════════════════════════════════════════════════
// Scene type (ESceneType)
// ═══════════════════════════════════════════════════════════════════════════════
export const SceneType = {
  Scene2D: 1,
  Scene3D: 2,
} as const;

// ═══════════════════════════════════════════════════════════════════════════════
// Server → Client message interfaces
// ═══════════════════════════════════════════════════════════════════════════════

/** Wire envelope: [type, payload] — wraps all messages with a type discriminator */
export interface MessageEnvelope {
  type: number;
  payload: Uint8Array;
}

/** SceneInitMsg: keys [0..4] — full scene snapshot sent on client connect */
export interface SceneInitMsg {
  sceneType: number;        // ESceneType
  bgColor: Uint8Array;      // RGBA bytes
  bounds: number[];          // 2D: [x0,y0,x1,y1], 3D: [x0,y0,z0,x1,y1,z1]
  transforms: number[];     // Flattened Mat4F array (16 floats per transform)
  entities: EntityDataMsg[];
}

/** EntityDataMsg: keys [0..1] — one entity's complete render data */
export interface EntityDataMsg {
  id: number;
  primitives: RenderPrimitiveMsg[];
}

/** RenderPrimitiveMsg: keys [0..14] — a single renderable primitive with draw attributes */
export interface RenderPrimitiveMsg {
  type: number;                                      // EPrimType (server enum, starts at 1)
  data: Float32Array | Float64Array | number[];      // Vertex data (floats)
  indices: Uint32Array | Int32Array | number[] | null; // Index data (optional, for Mesh3D triangles)
  color: Uint8Array;                                 // RGBA bytes
  lineWidth: number;
  lineType: number;                                  // ELineType
  ltScale: number;
  pointSize: number;
  transformIndex: number;
  zLevel: number;
  shadeMode: number;                                 // EShadeMode
  text: string | null;
  textAlign: number;                                 // ETextAlign
  wireIndices: Uint32Array | Int32Array | number[] | null; // Wire edge indices (separate from triangle indices)
  boundData: Float32Array | Float64Array | number[] | null; // Fill bounds [minX, minY, maxX, maxY]
}

/** EntityAddMsg: key [0] — entity added to scene */
export interface EntityAddMsg {
  entity: EntityDataMsg;
}

/** EntityRemoveMsg: key [0] — entity removed from scene */
export interface EntityRemoveMsg {
  entityId: number;
}

/** EntityUpdateMsg: key [0] — entity updated (full replacement of render data) */
export interface EntityUpdateMsg {
  entity: EntityDataMsg;
}

/** PickResultMsg: keys [0..1] — result of a pick operation */
export interface PickResultMsg {
  entityId: number;    // -1 if nothing picked
  position: number[];  // World position [x,y] or [x,y,z]
}

// ═══════════════════════════════════════════════════════════════════════════════
// Client → Server message interfaces
// ═══════════════════════════════════════════════════════════════════════════════

/** ViewStateMsg: keys [0..6] — client's current view state */
export interface ViewStateMsg {
  viewportW: number;
  viewportH: number;
  zoom: number;
  panX: number;
  panY: number;
  xRot: number;
  zRot: number;
}

/** PickRequestMsg: keys [0..1] — pick request from client (screen coordinates) */
export interface PickRequestMsg {
  x: number;
  y: number;
}

/** InteractionRequestMsg: keys [0..4] — interaction event from client */
export interface InteractionRequestMsg {
  type: number;       // EInteractionType
  x: number;          // World coordinates
  y: number;
  z: number;
  modifiers: number;  // Shift=1, Ctrl=2, Alt=4
}
